infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/KubernetesEnvironment.java

```diff
@@ -100,21 +132,72 @@ protected KubernetesEnvironment(
     super(internalRecipe, machines, warnings);
     setType(TYPE);
     this.pods = pods;
+    this.deployments = deployments;
     this.services = services;
     this.ingresses = ingresses;
     this.persistentVolumeClaims = persistentVolumeClaims;
     this.secrets = secrets;
     this.configMaps = configMaps;
+    this.podData = new HashMap<>();
+    pods.entrySet()
+        .forEach(
+            e -> {
+              Pod pod = e.getValue();
+              podData.put(e.getKey(), new PodData(pod.getSpec(), pod.getMetadata()));
+            });
+    deployments
+        .entrySet()
+        .forEach(
+            e -> {
+              PodTemplateSpec podTemplate = e.getValue().getSpec().getTemplate();
+              podData.put(
+                  e.getKey(), new PodData(podTemplate.getSpec(), podTemplate.getMetadata()));
+            });
   }
 
   @Override
   public KubernetesEnvironment setType(String type) {
     return (KubernetesEnvironment) super.setType(type);
   }
 
-  /** Returns pods that should be created when environment starts. */
-  public Map<String, Pod> getPods() {
-    return pods;
+  /**
+   * Returns pods that should be created when environment starts.
+   *
+   * <p>Note: This map <b>should not</b> be changed, as it will only return pods and not
+   * deployments. If objects in the map need to be changed, see {@link #getPodsData()}
+   */
+  public Map<String, Pod> getPodsCopy() {
+    return ImmutableMap.copyOf(pods);
+  }
+
+  /**
+   * Returns deployments that should be created when environment starts.
+   *
+   * <p>Note: This map <b>should not</b> be changed. If objects in the map need to be changed, see
+   * {@link #getPodsData()}
```

Comment:
This is tricky, there are kind of 3 levels of modification here.

- `#getPods()` and `#getDeployments()` cannot be modified at all, since you have no way of knowing if you have every deployable object in the workspace
- `#getPodsData()` *can* be modified, in the sense that it's what we'd use for provisioning. You cannot add or delete pods/deployments, but you can modify the objects in the map themselves
- `#addPod()` is the only way to change which deployable objects are in the environment.

I'll add some docs to clear this up.

Reply:
Added some docs, hopefully it's a little more clear.